checker.go

```diff
@@ -0,0 +1,229 @@
+// error, and reports whether the check succeeded.
+func (c *Checker) Check(f func(EqualFunc) error, a ...interface{}) bool {
+	if err := f(c.diff); err != nil {
+		c.fail(render(err, a))
+		return false
+	}
+	return true
+}
+
+func (c *Checker) diff(got, want interface{}) error {
+	if d := cmp.Diff(got, want, c.opts...); d != "" {
+		return fmt.Errorf("(-got +want)\n%s", d)
+	}
+	return nil
+}
+
+// Equal fails if the two provided values are not equal, and reports whether
+// the check succeeded.
+func (c *Checker) Equal(got, want interface{}, a ...interface{}) bool {
+	return c.Check(func(equal EqualFunc) error {
+		return equal(got, want)
+	}, a...)
+}
+
+// NotEqual fails if the two provided values are equal, and reports whether the
+// check succeeded.
+func (c *Checker) NotEqual(got, doNotWant interface{}, a ...interface{}) bool {
+	return c.Check(func(equal EqualFunc) error {
+		if err := equal(got, doNotWant); err != nil {
+			return nil
+		}
+		return fmt.Errorf("both values equal %q but should not", got)
+	}, a...)
+}
+
+// ErrorMatch fails if the given error is nil or does not match the given
+// regular expression pattern, and reports whether the check succeeded.
+func (c *Checker) ErrorMatch(got error, pattern string, a ...interface{}) bool {
+	return c.Check(ErrorMatch(got, pattern), a...)
+}
+
+// Panic fails if the given function does not panic or if the panic message
+// does not match the given pattern. It reports whether the check succeeded.
+func (c *Checker) Panic(f func(), pattern string, a ...interface{}) bool {
+	return c.Check(Panic(f, pattern), a...)
+}
+
+// EqualFunc is used to assert equality of got and want. An error is returned
+// if the provided values are not equal.
+type EqualFunc func(got, want interface{}) error
+
+// ErrorMatch returns a function suitable for being used with Checker.Check.
+// The returned function fails if the given error is nil or does not match the
+// given regular expression pattern.
+func ErrorMatch(got error, pattern string) func(EqualFunc) error {
+	return func(EqualFunc) error {
+		if got == nil {
+			return fmt.Errorf("error is nil, therefore it does not match %q\n", pattern)
+		}
+		return match(got, pattern)
+	}
+}
+
+// Panic returns a function suitable for being used with Checker.Check.
+// The returned function fails if the given function does not panic or if the
+// panic message does not match the given pattern.
+func Panic(f func(), pattern string) func(EqualFunc) error {
+	return func(EqualFunc) (err error) {
+		defer func() {
+			r := recover()
+			if r == nil {
+				err = fmt.Errorf("the function did not panic")
+				return
+			}
+			panicErr, ok := r.(error)
+			if !ok {
+				panicErr = fmt.Errorf("%s", r)
+			}
+			err = match(panicErr, pattern)
+		}()
+		f()
+		return nil
+	}
+}
+
+// NewNotEqualError returns an error that simplifies outputting "(-got +want)"
+// messages.
+func NewNotEqualError(msg, got, want string) error {
```

Comment:
This seems linked closely enough to the unexported render method that I'm not sure it helps to export it.